### PR TITLE
MAYA-113148 MayaUSD: Pass UsdMayaPrimReaderContext as reference inste…

### PR DESCRIPTION
**Pull request by boudrey**
…ad of pointer

### CODE REVIEW

File: lib/mayaUsd/fileio/primReader.h [hunk 1]
```diff
@@ -33,24 +33,22 @@ class UsdMayaPrimReader
 
     /// Reads the USD prim given by the prim reader args into a Maya shape,
     /// modifying the prim reader context as a result.
-    /// Callers must ensure \p context is non-null.
```
Comment (boudrey) [inline lib/mayaUsd/fileio/primReader.h:36]: Not necessary anymore.

File: lib/mayaUsd/fileio/shading/shadingModeUseRegistry.cpp [hunk 2]
```diff
@@ -602,7 +602,8 @@ class UseRegistryShadingModeImporter
     {
         // UsdMayaPrimReader::Read is a function that works by indirect effect. It will return
         // "true" on success, and the resulting changes will be found in the _context object.
-        if (!shaderReader.Read(_context->GetPrimReaderContext())) {
```
Comment (boudrey) [inline lib/mayaUsd/fileio/shading/shadingModeUseRegistry.cpp:605]: This is the only tricky place, I tried to reproduce the logic.

File: lib/mayaUsd/fileio/translators/translatorLight.cpp [hunk 3]
```diff
@@ -465,12 +465,12 @@ bool UsdMayaTranslatorLight::Read(
     }
 
     // Find which maya node needs to be our light's parent
-    MObject parentNode = context->GetMayaNode(lightSchema.GetPath().GetParentPath(), false);
+    MObject parentNode = context.GetMayaNode(lightSchema.GetPath().GetParentPath(), false);
```
Comment (boudrey) [inline lib/mayaUsd/fileio/translators/translatorLight.cpp:468]: Many places like here were using the pointer without testing it.

File: lib/usd/translators/meshReader.cpp [hunk 4]
```diff
@@ -67,12 +67,8 @@ TF_REGISTRY_FUNCTION_WITH_TAG(UsdMayaPrimReaderRegistry, UsdGeomMesh)
     });
 }
 
-bool MayaUsdPrimReaderMesh::Read(UsdMayaPrimReaderContext* context)
+bool MayaUsdPrimReaderMesh::Read(UsdMayaPrimReaderContext& context)
 {
-    if (!context) {
```
Comment (boudrey) [inline lib/usd/translators/meshReader.cpp:72]: Not necessary anymore.

**Comment by kxl-adsk** [comment]
@dj-mcg Can you help review this change?

**Review by JGamache-autodesk — APPROVED** [review]
Looks good to me. Any plans to propagate that change to callees like UsdMayaReadUtil::ReadUsdAttribute?

**Comment by dj-mcg** [comment]
Internally, Pixar likes to use pointers for output parameters to allow readers to immediately identify them. This isn't really an output parameter per se, and I wasn't able to find a site that might actually reassign the pointer itself, but it looks like that parameter can be mutated in the logic. I think this should be fine.

**Review by dj-mcg — APPROVED** [review]
Complies with Pixar's output-params-as-pointers as it's not exactly an output param. I think it looks good

**Comment by boudrey** [comment]
I looked at UsdMayaReadUtil::ReadUsdAttribute but decided not to extend the change to this function because of a condition in _ReadAnimatedUsdAttribute. "(context) ? context->GetTimeSampleMultiplier() : 1.0", context could be NULL here. I didn't know if this condition is used.

**Pull request merged**